src/TraverseTracks.cpp

```diff
@@ -0,0 +1,153 @@
+ *          during construction is used to redirect to the appropriate looping
+ *          scheme. If kernels are provided (not NULL) then they are deleted at
+ *          the end of the looping scheme.
+ * @param kernels MOCKernels to apply to all segments
+ */
+void TraverseTracks::loopOverTracksByParallelGroup(MOCKernel** kernels) {
+  loopOverTracksByParallelGroup2D(kernels);
+}
+
+
+/**
+ * @brief Loops over all explicit 2D Tracks
+ * @details The onTrack(...) function is applied to all 2D Tracks and the
+ *          specified kernels are applied to all segments. If NULL is provided
+ *          for the kernels, only the onTrack(...) functionality is applied.
+ * @param kernels The MOCKernels dictating the functionality to apply to
+ *        segments
+ */
+void TraverseTracks::loopOverTracks2D(MOCKernel** kernels) {
+
+  /* Loop over all parallel tracks for each azimuthal angle */
+  Track** tracks_2D = _track_generator->getTracks();
+  int num_azim = _track_generator->getNumAzim();
+  for (int a=0; a < num_azim/2; a++) {
+    int num_xy = _track_generator->getNumX(a) + _track_generator->getNumY(a);
+#pragma omp for
+    for (int i=0; i < num_xy; i++) {
+
+      Track* track_2D = &tracks_2D[a][i];
+
+      /* Operate on segments if necessary */
+      if (kernels != NULL) {
+        kernels[0]->newTrack(track_2D);
+        traceSegmentsExplicit(track_2D, kernels[0]);
+      }
+
+      /* Operate on the Track */
+      segment* segments = track_2D->getSegments();
+      onTrack(track_2D, segments);
+    }
+  }
+}
+
+
+/**
+ * @brief Loops over all explicit 2D Tracks by parallel group
+ * @details The onTrack(...) function is applied to all 2D Tracks and the
+ *          specified kernels are applied to all segments. If NULL is provided
+ *          for the kernels, only the onTrack(...) functionality is applied.
+ * @param kernels The MOCKernels dictating the functionality to apply to
+ *        segments
+ */
+void TraverseTracks::loopOverTracksByParallelGroup2D(MOCKernel** kernels) {
+
+  int min_track = 0;
+  int max_track = 0;
+
+  /* Loop over the parallel track groups */
+  Track** tracks = _track_generator->getTracksByParallelGroup();
+  int num_track_groups = _track_generator->getNumParallelTrackGroups();
+  for (int i=0; i < num_track_groups; i++) {
+
+    /* Compute the minimum and maximum Track IDs corresponding to
+     * this parallel track group */
+    min_track = max_track;
+    max_track += _track_generator->getNumTracksByParallelGroup(i);
+
+    /* Loop over each thread within this azimuthal angle halfspace */
+#pragma omp for schedule(guided)
+    for (int track_id=min_track; track_id < max_track; track_id++) {
+
+      Track* track = tracks[track_id];
+
+      /* Operate on segments if necessary */
+      if (kernels != NULL) {
+        kernels[0]->newTrack(track);
+        traceSegmentsExplicit(track, kernels[0]);
+      }
+
+      /* Operate on the Track */
+      segment* segments = track->getSegments();
```

Comment:
If you pass in a `Track` to the `onTrack(...)` method, why do you need to extract the segments from the `Track` as a separate parameter? Can't the `onTrack(...)` method simply extract the segments from the `Track` parameter itself? The segments seem like an unnecessary parameter.